### ISSUE TITLE
PLEASE COMMENT: noexcept specifier and set_terminate()

### ISSUE THREAD

**Issue by nobugs-hare** [O-Log-N/Hare-IDL#31]
In C++11, there is a new keyword noexcept, which is implemented as a runtime(!) check so that function declared as a noexcept, doesn't throw. Along the lines of "HAREASSERT everything, then HAREASSERT some more" I tend to like the idea, and suggest to discuss it here. 

One obstacle on going this way is that when a function declared with noexcept, throws, it just calls std::terminate(), so all the information on "which exception has caused the crash?" is lost (making it MUCH worse then asserts). OTOH, there _seems_ to be a workaround described in https://akrzemi1.wordpress.com/2011/10/05/using-stdterminate/ :

[[noreturn]] void onTerminate() noexcept
{
    if( auto exc = std::current_exception() ) { 
        // we have an exception
        try{
            rethrow_exception( exc ); // throw to recognize the type
        }
        catch( const std::exception& exc ) {
            cerr << exc.what();
        }
    }

```
std::_Exit( EXIT_FAILURE );
```

}

 In short - in case when function-declared-as-noexcept throws, std::terminate() is called, _but_ std::current_exception() _seems_ to work within std::terminate(), which allows to log the exception properly (to cerr/whatever) before terminating.

So, please comment (a) whether you like the idea, (b) if you do - whether above onTerminate() works with your compiler (reporting exception thrown from noexcept function, to console).

**Comment by marcosbracco** [comment]
In my MSVC 2015 it does not work. `std::current_exception()` returns `0` when called from terminate

Other issue with exceptions I just realized is at functions declared as `extern "C"`, which I have a lot at lex /yacc interaction. Compiler may assume they are `noexcept` depending on compiler flags. Will need to recheck this.

**Comment by marcosbracco** [comment]
Other option may be to always log a failed `HAREASSERT` to `stderr` **before** throwing.

In any real situation where there is an upper piece of code that can catch the `AssertException` and make something (like retry the operation, or something like that), it won't hurt much to have an unwanted message at `stderr`.
In the current situation where we are just giving up, we will get the same behavior as we have now.

**Comment by nobugs-hare** [comment]
HAREASSERT is easy, but I see the following (and pretty BIG) problem with noexcept: if something (not "our" code) causes an exception which runs into _noexcept_ function - how do we know what happened? This lack-of-tracing for 3rd-party exceptions tends to make me discourage _noexcept_ :-(

**Comment by marcosbracco** [comment]
This one seems not to have easy solution.
It seems that _noexcept_ is more about compiler optimization than about exception specifications or static code checks.
My current opinion is to default to **avoid** _noexcept_, and **only** use it at _move constructor_ (after careful review) where it seems to be the most important use.
What do you think?

**Comment by nobugs-hare** [comment]
Yes, I tend to agree (it is a pity, but there seems to be no good workaround in modern C++ to make noexcept semantically meaningful).
